### ISSUE TITLE
Interpolation comment tagging API proposal

### ISSUE THREAD

**Issue by emilgoldsmith** [styled-components/stylelint-processor-styled-components#87]
To take the next step for #70 that followed from #68 I thought I would just put my initial thoughts on an API for this new feature. I thought it would be better to keep it separate here instead of in comments in the PR or something for nice organization if it ends up becoming a longer discussion.

So first of all we of course need a prefix for the comments like there is in `stylelint-enable`, `eslint-disable` etc. where I would simply suggest `scp` for Styled Components Processor but still keeping it very short as one of my considerations for this API is that we want to keep it as short as possible as for a short interpolation such as `color: ${props => props.theme.color}` you don't want to add a comment that's longer than the interpolation itself making inline short interpolations inconvenient.

I'll keep using `scp` as the prefix for my following suggestions but that is of course also up for discussion.

Following is my current draft of the full api (all vocabulary taken from [here](http://apps.workflower.fi/vocabs/css/en#statement)):
- `scp-ref` (I'm thinking this should work for cases like `${Button} { css.... }`
- `scp-block`
- `scp-sel` (sel is short for selector)
- `scp-dec` (dec short for declaration)
- `scp-prop` (prop short for property)
- `scp-val` (val short for value)
- `scp-custom PLACEHOLDER` (PLACEHOLDER is what we will replace the interpolation with for linting purposes)

This is just my quick first thoughts, and it's of course also easy to extend this API or modify it later.

Some of my main considerations are:
- Should the prefix be different (though I'm personally pretty happy with `scp`)
- Should some of the API be shortened further / shortened less (such as using `value` instead of `val` or `cust` instead of `custom` etc.)
- What should the exact API for `scp-custom` be? My initial thought is to give maximum customizeability to the user so to literally say we don't count the one space between `scp-custom` and `PLACEHOLDER` but after that we literally take whatever is left. The question is then should you trim one final space at the end? Should you trim completely the end? You could also trim the beginning though that is what I was thinking not to do to give max customizeability. One could of course also make `scp-custom` be trimmed in both ends of `PLACEHOLDER` while you could have like a super custom like `scp-all-cust` or w/e that didn't trim as the not trimming could of course also be a nuisance to users as a payoff for the customizeability. 
- How specific should we make the API? For example it is very likely that we would in the implementation use the same placeholder for `scp-block` and `scp-dec` as a block and a declaration seems the same, but I thought it might be nice UX and even coding style (as in if someone else is reading your code they know that whatever function you put in will be inputting a block?) if you could had the possibility to be more specific even if the actual implementation might not distinguish between some of them.

I may have forgotten something but that's what the discussion is for :). And I may do an edit to this post depending on how the discussion evolves below here with an updated API proposal as things change.

**MOST UPDATED PROPOSAL:**
- `sc-ref`
- `sc-block`
- `sc-selector`
- `sc-declaration`
- `sc-property`
- `sc-value`
- `sc-custom PLACEHOLDER`

And allow unique prefixes to be used to give flexibility which in the current case would even mean that `sc-b` would mean `sc-block`, it would be at the developers discretion to make it readable.

**Comment by mxstbr** [comment]
What about using just `sc-` as the prefix? I think most people don't care that it's the _processor_ they're talking to, they really just want to mark a `styled-components` code snippet as something semantically.

In my mind from a user perspective `sc-sel` makes more sense than `scp-sel`?

**Comment by mxstbr** [comment]
I think we should allow both long and short versions, then people can choose themselves if they value clarity or terseness depending on their team and project, it seems weird that we would choose that categorically for all users.

That being said, we then need to have documentation about this feature (btw: we should probably move the docs for the processor to the styled-components website) and we'll need to choose which style to go for in the docs. I don't know what the choice there is, likely the long, readable version and a note that you can also use shorthands?

**Comment by emilgoldsmith** [comment]
> What about using just sc- as the prefix? I think most people don't care that it's the processor they're talking to, they really just want to mark a styled-components code snippet as something semantically.

Completely agree with that, no need to separate ourselves from Styled Components at all.

> I think we should allow both long and short versions, then people can choose themselves if they value clarity or terseness depending on their team and project,

Yep good point. I think I remember talking about something like that in #68 where I proposed writing a bit of code that allows the user to just trim as much as you want as long as it's a unique prefix it'll then find the one you're looking for, a bit like autocomplete. That would also allow flexible shortenings and shouldn't be too hard.

> I don't know what the choice there is, likely the long, readable version and a note that you can also use shorthands?

Yep I like that too

**Comment by emilgoldsmith** [comment]
I updated the proposal above. I guess an argument could be made for it being ugly even allowing for something like `sc-b` as mentioned in the update, even if it is up to the developer. What do you guys think? Flexibility or forced good style by deciding on the short names?

Also, what do you guys think on the questions I mentioned regarding the `scp-custom` and where to trim / not to trim.

**Comment by mxstbr** [comment]
> allows the user to just trim as much as you want as long as it's a unique prefix

That's a pretty cool idea, I like it! `tmux` does that with their commands and it's very useful. It doesn't need to be like super well documented or anything but just like use the longhand in the docs and then have a note that says you can shorten them.

If that's too complex in terms of the code I'd also be fine with just having a map of shorthands to map `val` to `value`.

----

What exactly is the difference between `sc-ref` and `sc-sel`? A ref gets replaced by a selector anyway, so why would we have a custom type for that?

----

I think with `sc-custom` it makes sense to trim based on the way they write their comment? That's pretty unambiguous but I don't know if we have the data. This is what I imagine:

```JS
// Simple cases:
/* sc-custom hello world */ // => "hello world"
/*sc-custom hello world*/ // => "hello world"

// Edge cases:
/* sc-custom hello world*/ // => "hello world"
/*sc-custom hello world */ // => "hello world "
```

Basically, if they include a space in the way they write comments we'll trim it, if they don't have spaces in their comments but then add one at the end we don't trim it?

**Comment by emilgoldsmith** [comment]
> If that's too complex in terms of the code I'd also be fine with just having a map of shorthands to map val to value.

I don't think it'll be too bad, I think we could do it pretty simply, I'll try and see, if it gets out of hand I'll just revert to the simpler.

----

> What exactly is the difference between sc-ref and sc-sel?

Hmm, I guess the reason I added it was because it was one of the things way in the front of my mind because a lot of the edge cases we want to fix are related to refs. Now that I think about it I kinda see your point, I guess it's not exactly the same as `sc-block` and `sc-declaration` which are more semantically different than `ref` and `selector` I guess. I'm not quite sure which one I prefer (removing `ref` or not), but the argument for keeping it would be that it leaves more room for better documentation in terms of code maintainability or something I guess by being more specific in your interpolation tagging. But `selector` may very well be more than good enough.

----

> Basically, if they include a space in the way they write comments we'll trim it, if they don't have spaces in their comments but then add one at the end we don't trim it?

Hmm... interesting. We'd have to document that for sure, and then it would be okay, I just don't feel like that's so intuitive. As sometimes you really would need the space, and I feel like it's very unintuitive that 
`/*sc-custom hello world */` and `/* sc-custom hello world */` are different and not an inconsistency, ESLint or prettier could also possibly complain?

What about the following in pseudo code, which is basically what you said with a bit more detail and nuance:
1. If there are no newlines in the start or end of the comment (I guess mostly there never will be but it's an edge case we should consider) we take whatever whitespace that is in the start of the comment and at minimum subtract that from the end (if there's less whitespace at the end remove whatever there is).
2. If there are newlines in the start or end we simply trim.

We should probably have some slight documentation for that as well, but I feel like this is more intuitive and it would allow the distinction to now be `/* sc-custom hello world */` and `/* sc-custom hello world  /*` if your custom is to put 1 space in start and end. What do you think? Maybe a different approach to 2.?

**Comment by ismay** [comment]
This might be a stupid question, but why not use single or double quotes as delimiters: `/* sc-custom 'hello world' */` or `/* sc-custom 'hello world  ' */`? Would that not be clearer in terms of parsing the placeholder comments?

**Comment by emilgoldsmith** [comment]
> This might be a stupid question, but why not use single or double quotes as delimiters: /* sc-custom 'hello world' */ or /* sc-custom 'hello world ' */? Would that not be clearer in terms of parsing the placeholder comments?

I don't find that a stupid question at all! I love that idea, let's do it! It's a really clear API and makes our coding job simple and unambiguous :). I'm sold.

**Comment by mxstbr** [comment]
Done in #70! (not closed because it was merged into `v1` rather than `master`, so manually closing)
